lib/internal/modules/esm/resolve.js

```diff
@@ -938,19 +948,77 @@ function defaultResolve(specifier, context = {}, defaultResolveUnused) {
   let parsed;
   try {
     parsed = new URL(specifier);
-    if (parsed.protocol === 'data:') {
+    if (!parentURL) {
+      // allow
+    }
+    if (parsed.protocol === 'data:' ||
+        (experimentalHttpsModules && (
+          parsed.protocol === 'https:' ||
+          parsed.protocol === 'http:'))) {
       return {
-        url: specifier
+        url: specifier,
       };
     }
-  } catch {}
-  if (parsed && parsed.protocol === 'node:')
-    return { url: specifier };
-  if (parsed && parsed.protocol !== 'file:' && parsed.protocol !== 'data:')
-    throw new ERR_UNSUPPORTED_ESM_URL_SCHEME(parsed);
+  } catch { }
+  if (parentURL) {
+    const parsedParentURL = new URL(parentURL);
+    if (parsedParentURL.protocol === 'http:' ||
+        parsedParentURL.protocol === 'https:') {
+      if (shouldBeTreatedAsRelativeOrAbsolutePath(specifier)) {
+        parsed = new URL(specifier, parsedParentURL);
+        // data: and blob: disallowed due to allowing file: access via
+        // indirection
+        if (parsed &&
+            parsed.protocol !== 'https:' &&
+            parsed.protocol !== 'http:') {
+          throw new ERR_UNSUPPORTED_ESM_NETWORK_IMPORT_SPECIFIER(
+            specifier,
+            parentURL,
+            'cannot import local location from remote location,' +
+            'load a it via a remote location.');
+        } else {
+          return { url: parsed.href };
+        }
+      } else {
+        throw new ERR_UNSUPPORTED_ESM_NETWORK_IMPORT_SPECIFIER(
+          specifier,
+          parentURL,
+          'only relative and absolute specifiers are supported.');
+      }
+    }
+  }
+  if (parsed && parsed.protocol === 'node:') return { url: specifier };
+  if (
+    parsed &&
+    parsed.protocol !== 'file:' &&
+    parsed.protocol !== 'data:' &&
+    (
+      !experimentalHttpsModules || (
+        parsed.protocol !== 'https:' &&
+        parsed.protocol !== 'http:')
+    )
+  )
+    throw new ERR_UNSUPPORTED_ESM_URL_SCHEME(parsed, [
+      'file',
+      'data',
+      ...(experimentalHttpsModules ? ['https', 'http'] : []),
+    ]);
   if (NativeModule.canBeRequiredByUsers(specifier)) {
+    if (parentURL) {
+      const parsedParentURL = new URL(parentURL);
+      if (
+        parsedParentURL.protocol === 'http:' ||
+        parsedParentURL.protocol === 'https:'
+      ) {
+        throw new ERR_UNSUPPORTED_ESM_NETWORK_IMPORT_SPECIFIER(
+          specifier,
+          parentURL,
+          'cannot import local location from remote location,' +
+          'load a it via a remote location.');
+      }
+    }
```

Comment:
For example, we could just replace this whole block with a separate `throwUnsupportedIfRemoteParent(parentURL)` function.

Reply:
seems doable